CMSgov/beneficiary-fhir-data
# BFD-3420: Relax claim endpoint id validation regex pattern

**JIRA Ticket:**
[BFD-3420](https://jira.cms.gov/browse/BFD-3420)

**User Story or Bug Summary:**

What:

The claim endpoint validates that the claim ID supplied in the URL only consists of numbers
This is incorrect because the claim ID for FISS claims is alphanumeric
Why:

The endpoint is currently unusable due to this issue
This hasn't been noticed in the past because this endpoint isn't actually used in prod, but it's useful for internal testing
Desired Outcome:

Fix the endpoint so it can parse the FISS claim ID
---

### What Does This PR Do?


Change CLAIM_ID_PATTERN regex FROM: "([fm])-(-?\\p{Digit}+)"  TO: "([fm])-(-?[A-Za-z0-9]+)"

### What Should Reviewers Watch For?


If you're reviewing this PR, please check for these things in particular:

* Verify all PR security questions and checklists have been completed and addressed.
* Locally passing all tests
* Manually test fiss claim get, and mcs claim get as below:

Manual test steps:

1. Check out PR to local repo
2. Follow instruction to start a local bfd server with -z
3. Run IT to populate a fiss claim (see CLI example)
4. curl the claim endpoint with fiss claim ID (see CLI example), and exam the claim json response.
5. curl the claim endpoint with mcs claim ID (see CLI example), and exam the claim json response.

Example CLIs and responses:

1. Create fiss claim record and fetch it using curl/postman, run below commands:

`
cd ./apps/bfd-pipeline/bfd-pipeline-rda-grpc

mvnDebug -Dits.db.url="jdbc:postgresql://localhost:5432/bfd" -Dits.db.username=fhirdb -Dits.db.password=InsecureLocalDev -Dmaven.build.cache.enabled=false -DforkCount=0 -Dit.test=FissClaimRdaSinkIT#fissClaim clean verify

Attach you IDE to 8000, break at FissClaimRdaSinkIT line 146, where fiss claim record populated (and data not tear down yet)

curl --silent --insecure --cert $BFD_PATH/apps/bfd-server/dev/ssl-stores/client-unsecured.pem "https://localhost:6500/v2/fhir/Claim/f-MW1k"

`

2. Create mcs claim record and fetch it, run below commands:

`
cd ./apps/bfd-pipeline/bfd-pipeline-rda-grpc

mvnDebug -Dits.db.url="jdbc:postgresql://localhost:5432/bfd" -Dits.db.username=fhirdb -Dits.db.password=InsecureLocalDev -Dmaven.build.cache.enabled=false -DforkCount=0 -Dit.test=McsClaimRdaSinkIT#mcsClaim clean verify

Attach you IDE to 8000, break at McsClaimRdaSinkIT line 129, where mcs claim record populated (and data not tear down yet)

curl --silent --insecure --cert $BFD_PATH/apps/bfd-server/dev/ssl-stores/client-unsecured.pem "https://localhost:6500/v2/fhir/Claim/m-3"

`

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
    * Does this PR modify or invalidate any of our security controls?
    * Does this PR store or transmit data that was not stored or transmitted before?

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?

### What Needs to Be Merged and Deployed Before this PR?



This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* N/A


### Submitter Checklist


I have gone through and verified that...:

* [x] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    6. There isn't too much of a burden on reviewers.
    7. Any problems it causes have a small "blast radius".
    8. It'll be easier to rollback if that becomes necessary.
* [ ] The data dictionary has been updated with any field mapping changes, if any were made.
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
    * Please review the standards every few months to ensure you're familiar with them.
